### PR TITLE
Add VDF

### PR DESCRIPTION
**Pull request by jonas-lj**
Add an implementation of a VDF using [Wesolowski's](https://eprint.iacr.org/2018/623.pdf) construction with imaginary class groups. The implementation is compatible with chiavdf's library and has been tested against test vectors produced using using their library.

The group is generic and RSA groups may be added in a future PR.

### CODE REVIEW

**Review by kchalkias — COMMENTED** [review]
@jonas-lj can we have a quick presentation on the algorithm next week. It will help us going through details for the code review

**Comment by jonas-lj** [comment]
> @jonas-lj can we have a quick presentation on the algorithm next week. It will help us going through details for the code review

Sure. I'm also making some changes so don't spend too much time on it yet.

File: fastcrypto/src/vdf.rs [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        iterations: u64,
+    ) -> FastCryptoResult<bool>;
+}
+
+/// An implementation of the Wesolowski VDF construction over ideal class groups. The implementation
+/// is compatible with chiavdf.
+///
+/// Note that the evaluation phase is currently significantly slower than other implementations, so
+/// estimates on how long it takes to evaluate a VDF should currently be used cautiously.
+pub struct ClassgroupVDF {
```
Comment (benr-ml) [inline fastcrypto/src/vdf.rs:48]:
```suggestion
pub struct ClassGroupVDF {
```

File: fastcrypto/src/vdf.rs [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+/// This represents a Verifiable Delay Function (VDF) construction over a given group.
+pub trait VDF {
+    type GroupElement;
+
+    /// Evaluate this VDF and return the output and proof of correctness.
+    fn prove(
+        &self,
+        input: &Self::GroupElement,
+        iterations: u64,
+    ) -> FastCryptoResult<(Self::GroupElement, Self::GroupElement)>;
```
[the remark this answers is not in the document]
Comment (jonas-lj) [reply fastcrypto/src/vdf.rs:39]: Good point. I've made the VDF trait more generic and moved the specifics to the WesolowskiVDF type.

File: fastcrypto/src/groups/class_group/mod.rs [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+        self.compose(self)
+    }
+
+    fn mul(&self, scale: &BigInt) -> Self {
+        if scale.is_zero() {
+            return Self::zero(&self.discriminant());
+        }
+
+        let mut result = self.clone();
+        for i in (0..scale.bits() - 1).rev() {
```
Comment (benr-ml) [inline fastcrypto/src/groups/class_group/mod.rs:266]: there is no iterator on the bits?

File: fastcrypto/src/vdf/wesolowski.rs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+            return Err(InvalidInput);
+        }
+
+        if self.iterations == 0 {
+            return Ok((input.clone(), G::zero(&self.group_parameter)));
+        }
+
+        let mut output = input.clone();
+        for _ in 0..self.iterations {
+            output = output.clone() + &output;
```
Comment (benr-ml) [inline fastcrypto/src/vdf/wesolowski.rs:55]: double()?

File: fastcrypto/src/vdf/wesolowski.rs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+        if !input.has_group_parameter(&self.group_parameter) {
+            return Err(InvalidInput);
+        }
+
+        if self.iterations == 0 {
+            return Ok((input.clone(), G::zero(&self.group_parameter)));
+        }
+
+        let mut output = input.clone();
+        for _ in 0..self.iterations {
```
Comment (benr-ml) [inline fastcrypto/src/vdf/wesolowski.rs:54]: iterations = T when in the protocol it says 2^T?

File: fastcrypto/src/groups/mod.rs [hunk 6]
```diff
@@ -71,3 +72,33 @@ pub trait HashToGroupElement {
...
+    fn double(&self) -> Self;
+
+    /// Compute scale * self.
+    fn mul(&self, scale: &Self::ScalarType) -> Self;
+
+    /// Serialize this group element.
+    fn as_bytes(&self) -> Vec<u8>;
+
+    /// Check whether this group element is a member of a group with a given parameter.
+    fn has_group_parameter(&self, parameter: &Self::ParameterType) -> bool;
```
Comment (benr-ml) [inline fastcrypto/src/groups/mod.rs:100]: maybe `same_group(...)`?

File: fastcrypto/src/vdf/wesolowski.rs [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+
+    fn verify(&self, input: &G, output: &G, proof: &G) -> FastCryptoResult<()> {
+        if !input.has_group_parameter(&self.group_parameter)
+            || !output.has_group_parameter(&self.group_parameter)
+            || !proof.has_group_parameter(&self.group_parameter)
+        {
+            return Err(InvalidInput);
+        }
+
+        let challenge = get_challenge(&input.as_bytes(), &output.as_bytes());
```
Comment (benr-ml) [inline fastcrypto/src/vdf/wesolowski.rs:80]: doesn't it include also the group parameters?

**Review by benr-ml — APPROVED** [review]

File: fastcrypto/src/groups/class_group/mod.rs (hunk 3 above)
Comment (jonas-lj) [reply fastcrypto/src/groups/class_group/mod.rs:266]: No, there's only the bit length and the check bit functions.

File: fastcrypto/src/vdf/wesolowski.rs (hunk 5 above)
Comment (jonas-lj) [reply fastcrypto/src/vdf/wesolowski.rs:54]: Yes, exactly. It's documented in the constructor for the Wesolowski construction, but I'll add a comment here also.

File: fastcrypto/src/groups/mod.rs (hunk 6 above)
Comment (jonas-lj) [reply fastcrypto/src/groups/mod.rs:100]: Good idea!

File: fastcrypto/src/vdf/wesolowski.rs (hunk 7 above)
Comment (jonas-lj) [reply fastcrypto/src/vdf/wesolowski.rs:80]: No, but it could include a bit_length parameter which is right now given as a constant.

**Pull request merged**